showcase-v1alpha2/src/v1alpha2/echo_client.ts

```diff
@@ -0,0 +1,688 @@
+          this._descriptors.page[methodName] ||
+              this._descriptors.stream[methodName]);
+    }
+  }
+
+  /**
+   * The DNS address for this API service.
+   */
+  static get servicePath() {
+    return 'localhost';
+  }
+
+  /**
+   * The port for this API service.
+   */
+  static get port() {
+    return 7469;
+  }
+
+  /**
+   * The scopes needed to make gRPC calls for every method defined
+   * in this service.
+   */
+  static get scopes() {
+    return [
+      'https://www.googleapis.com/auth/cloud-platform',
+    ];
+  }
+
+  /**
+   * Return the project ID used by this class.
+   * @param {function(Error, string)} callback - the callback to
+   *   be called with the current project Id.
+   */
+  getProjectId(): Promise<string>;
+  getProjectId(callback: GapicCallback<string>): void;
+  getProjectId(callback?: GapicCallback<string>): Promise<string>|void {
+    if (!callback) {
+      return this.auth.getProjectId();
+    }
+    this.auth.getProjectId(callback);
+  }
+
+  // -------------------
+  // -- Service calls --
+  // -------------------
+
+  /**
+   * This method simply echos the request. This method is showcases unary rpcs.
+   *
+   * @param {Object} request
+   *   The request object that will be sent.
+   * @param {string} [request.content]
+   *   The content to be echoed by the server.
+   * @param {Object} [request.error]
+   *   The error to be thrown by the server.
+   *
+   *   This object should have the same structure as [Status]{@link
+   * google.rpc.Status}
+   * @param {Object} [options]
+   *   Optional parameters. You can override the default settings for this call,
+   * e.g, timeout, retries, paginations, etc. See [gax.CallOptions]{@link
+   * https://googleapis.github.io/gax-nodejs/global.html#CallOptions} for the
+   * details.
+   * @param {function(?Error, ?Object)} [callback]
+   *   The function which will be called with the result of the API call.
+   *
+   *   The second parameter to the callback is an object representing
+   * [EchoResponse]{@link google.showcase.v1alpha2.EchoResponse}.
+   * @returns {Promise} - The promise which resolves to an array.
+   *   The first element of the array is an object representing
+   * [EchoResponse]{@link google.showcase.v1alpha2.EchoResponse}. The promise
+   * has a method named "cancel" which cancels the ongoing API call.
+   *
+   * @example
+   *
+   * const showcase = require('showcase.v1alpha2');
+   *
+   * const client = new showcase.v1alpha2.EchoClient({
+   *   // optional auth parameters.
+   * });
+   *
+   *
+   * client.echo({})
+   *   .then(responses => {
+   *     const response = responses[0];
+   *     // doThingsWith(response)
+   *   })
+   *   .catch(err => {
+   *     console.error(err);
+   *   });
+   */
+  echo(request: EchoRequest, options?: gax.CallOptions): Promise<EchoResponse>;
+  echo(
+      request: EchoRequest, options: gax.CallOptions,
+      callback: GapicCallback<EchoResponse>): void;
+  echo(
```

Comment:
Perhaps these two could be...

```typescript
echo(request: EchoRequest, options?: gax.CallOptions,
    callback: GapicCallback<EchoResponse>): void;
```

Is that legal?

Reply:
Unfortunately, it's not, pretty much like in other languages: required parameters cannot go after optional parameters. Example from playground [here](http://www.typescriptlang.org/play/#src=interface%20RequestType%20%7B%0D%0A%20%20%20%20field%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20RequestOptions%20%7B%0D%0A%20%20%20%20setting%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Callback%20%7B%0D%0A%20%20%20%20(result%3A%20string)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Afunction%20test(request%3A%20RequestType%2C%0D%0A%20%20%20%20options%3F%3A%20RequestOptions)%3A%20Promise%3Cstring%3E%3B%0D%0Afunction%20test(request%3A%20RequestType%2C%0D%0A%2F%2F%20try%20this%3A%0D%0A%2F%2F%20%20options%3F%3A%20RequestOptions%2C%0D%0A%20%20%20%20options%3A%20RequestOptions%2C%0D%0A%20%20%20%20callback%3A%20Callback)%3A%20void%3B%0D%0Afunction%20test(request%3A%20RequestType%2C%0D%0A%20%20%20%20options%3F%3A%20RequestOptions%2C%0D%0A%20%20%20%20callback%3F%3A%20Callback)%3A%20Promise%3Cstring%3E%20%7C%20void%0D%0A%7B%0D%0A%20%20%20%20%2F%2F%20take%20care%20of%20default%20options%0D%0A%20%20%20%20options%20%3D%20Object.assign(%7B%20setting%3A%20%22no%20options%22%20%7D%2C%20options)%3B%0D%0A%20%20%20%20const%20result%20%3D%20%60fld%3A%20%24%7Brequest.field%7D%2C%20opt%3A%20%24%7Boptions.setting%7D%60%3B%0D%0A%20%20%20%20if%20(!callback)%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20new%20Promise(resolve%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20resolve(result)%3B%0D%0A%20%20%20%20%20%20%20%20%7D)%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20callback(result)%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20calling%20different%20signatures%0D%0A%0D%0Atest(%7B%20field%3A%20%22value%22%20%7D).then(alert)%3B%0D%0A%0D%0Atest(%7B%20field%3A%20%22value%22%20%7D%2C%20%7B%20setting%3A%20%22with%20options%22%20%7D).then(alert)%3B%0D%0A%0D%0A%2F%2F%20the%20following%20call%20does%20not%20compile%20without%20passing%20options%0D%0A%2F%2Ftest(%7B%20field%3A%20%22value%22%20%7D%2C%20alert)%3B%0D%0A%0D%0Atest(%7B%20field%3A%20%22value%22%20%7D%2C%20%7B%7D%2C%20alert)%3B%0D%0A). @JustinBeckwith can have more ideas here.

Reply:
Upd: thanks @JustinBeckwith, it's possible ([playground link](http://www.typescriptlang.org/play/#src=interface%20RequestType%20%7B%0D%0A%20%20%20%20field%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20RequestOptions%20%7B%0D%0A%20%20%20%20setting%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Callback%20%7B%0D%0A%20%20%20%20(result%3A%20string)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Afunction%20test(request%3A%20RequestType%2C%0D%0A%20%20%20%20options%3F%3A%20RequestOptions)%3A%20Promise%3Cstring%3E%3B%0D%0Afunction%20test(request%3A%20RequestType%2C%0D%0A%20%20%20%20callback%3A%20Callback)%3A%20void%3B%0D%0Afunction%20test(request%3A%20RequestType%2C%0D%0A%20%20%20%20options%3A%20RequestOptions%2C%0D%0A%20%20%20%20callback%3A%20Callback)%3A%20void%3B%0D%0Afunction%20test(request%3A%20RequestType%2C%0D%0A%20%20%20%20optionsOrCallback%3F%3A%20RequestOptions%7CCallback%2C%0D%0A%20%20%20%20callback%3F%3A%20Callback)%3A%20Promise%3Cstring%3E%20%7C%20void%0D%0A%7B%0D%0A%20%20%20%20let%20options%3A%20RequestOptions%20%3D%20%7B%7D%3B%0D%0A%20%20%20%20if%20(typeof%20optionsOrCallback%20%3D%3D%3D%20'function')%20%7B%0D%0A%20%20%20%20%20%20%20%20callback%20%3D%20optionsOrCallback%3B%0D%0A%20%20%20%20%7D%20else%20%7B%0D%0A%20%20%20%20%20%20%20%20options%20%3D%20optionsOrCallback%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20%2F%2F%20take%20care%20of%20default%20options%0D%0A%20%20%20%20options%20%3D%20Object.assign(%7B%20setting%3A%20%22no%20options%22%20%7D%2C%20options)%3B%0D%0A%20%20%20%20const%20result%20%3D%20%60fld%3A%20%24%7Brequest.field%7D%2C%20opt%3A%20%24%7Boptions.setting%7D%60%3B%0D%0A%20%20%20%20if%20(!callback)%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20new%20Promise(resolve%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20resolve(result)%3B%0D%0A%20%20%20%20%20%20%20%20%7D)%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20callback(result)%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20calling%20different%20signatures%0D%0A%0D%0Atest(%7B%20field%3A%20%22value%22%20%7D).then(alert)%3B%0D%0A%0D%0Atest(%7B%20field%3A%20%22value%22%20%7D%2C%20%7B%20setting%3A%20%22with%20options%22%20%7D).then(alert)%3B%0D%0A%0D%0Atest(%7B%20field%3A%20%22value%22%20%7D%2C%20alert)%3B%0D%0A%0D%0Atest(%7B%20field%3A%20%22value%22%20%7D%2C%20%7B%7D%2C%20alert)%3B%0D%0A)).